Doc/regions.md

```diff
@@ -0,0 +1,567 @@
+However, we need to support local objects becoming immutable to allow them to be shared, and suppport for groups of objects to be moved between interpreters.
+
+To pass mutable states between threads, we need to introduce a new construct to Python: "regions".
+A region is a group of objects that can move together across threads
+and can be accessed by at most one interpreter at a time.
+This removes the need for locks to protect the objects in the region.
+The region can be passed between threads, and the receiving interpreter can access the objects in the region.
+
+To efficiently support regions, we impose the following invariants that the runtime is responsible for maintaining:
+
+* Local objects can reach anything
+* Immutable objects can only reach other immutable objects.
+* Objects in a region can only reach other objects in the same region, immutable objects, or linear region entry points.
+* Regions track the number of incoming references from local objects.
+
+New objects are always created as a local object.
+If Veronapy features are not used, then all the objects in the system will be local objects and the runtime will behave as normal.
+
+
+### Regions
+
+Regions are used to pass mutable state between concurrent units of execution (behaviours).
+For usability, we need references into the region, that are not part of the region itself.
+These might be from the objects such as the frame, from closures or from function objects.
+This is precisely what local objects are for as they can refer into the region.
+
+The core challenge is allowing local objects to refer into a region,
+while allowing the region to be sent to another interpreter.
+To do this, we need to track the number of references from local objects to a region, the "local reference count" (LRC).
+
+When we send a region to another behaviour we need to ensure that the LRC is zero.
+To make this more usable, if the LRC is not zero, we can scan the current frame to find all references to the region, and remove them.
+If there are still references, then we can scan all objects (from the existing doubly-linked list in the Python cycle detector) to find a reference,
+and report that as an error.
+
+When a local object becomes reachable from a region, we need to promote it and anything it can reach to be part of that region.
+When we promote the local object we need to modify the LRC of the region.
+This involves adding new counts for reference to the newly added objects, and removing counts for edges that have become internal to the region.
+Consider the following example:
+
+```python
+x = {}
+#  x is a field of the frame object, which is local. 
+#  it points at a fresh dictionary object, which is also local.
+y = {}
+#  y is a field of the frame object, which is local. 
+#  it points at a fresh dictionary object, which is also local.
+x.f = y
+# x.f is a field of the dictionary object, which is local.
```

Comment:
Nit: in Python (as opposed to JavaScript, say), dictionaries don't have mutable attributes. A realistic example would perhaps use
```py
x["f"] = y
# x["f"] is a value slot in the dictionary object, which is local
```

Reply:
6be0539d14c1feb5be10898facfd3fe8ad0973fb